okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/FixedLengthSource.kt

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2021 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio.zipfilesystem
+
+import okio.Buffer
+import okio.ForwardingSource
+import okio.IOException
+import okio.Source
+
+/**
+ * A source that returns [size] bytes of [delegate].
+ *
+ * This throws an [IOException] if the delegate returns fewer than [size] bytes.
+ *
+ * If [truncate] is true, this truncates to [size] bytes. Otherwise this requires that [delegate]
+ * will return exactly [size] bytes, and will throw an [IOException] if it doesn't.
+ */
+internal class FixedLengthSource(
+  delegate: Source,
+  private val size: Long,
+  private val truncate: Boolean
+) : ForwardingSource(delegate) {
+  private var bytesReceived = 0L
+
+  override fun read(sink: Buffer, byteCount: Long): Long {
+    // Figure out how many bytes to attempt to read.
+    //
+    // If we're truncating, we never attempt to read more than what's remaining.
+    //
+    // Otherwise we expect the underlying source to be exactly the promised size. Read as much as
+    // possible and throw an exception if too many bytes are returned.
+    val toRead = when {
+      bytesReceived > size -> 0L // Already read more than the promised size.
```

Comment:
Why isn't this an error? Doesn't this state turn a presumed while loop in a caller into an infinite loop? e.g. If there are multiple concurrent threads calling this class?